src/mpid/ch4/generic/am/mpidig_am_msg.h

```diff
@@ -163,18 +179,27 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_recv_copy_seg(void *payload, MPI_Aint payloa
     MPIDIG_rreq_async_t *p = &(MPIDIG_REQUEST(rreq, req->async));
 
     if (p->recv_type == MPIDIG_RECV_DATATYPE) {
+        MPI_Aint max_data_size;
+        MPIR_Datatype_get_size_macro(MPIDIG_REQUEST(rreq, datatype), max_data_size);
+        max_data_size *= MPIDIG_REQUEST(rreq, count);
+
         MPI_Aint actual_unpack_bytes;
         MPIR_Typerep_unpack(payload, payload_sz,
                             MPIDIG_REQUEST(rreq, buffer),
                             MPIDIG_REQUEST(rreq, count),
                             MPIDIG_REQUEST(rreq, datatype), p->offset, &actual_unpack_bytes);
         p->in_data_sz -= actual_unpack_bytes;
         p->offset += actual_unpack_bytes;
-        if (payload_sz > actual_unpack_bytes) {
+        if (payload_sz + p->offset - actual_unpack_bytes > max_data_size) {
```

Comment:
I don't think this condition is correct or does what you mean. `payload_sz + p->offset` is the total amount of data received so far, right? Why subtract `actual_unpack_bytes`?

Reply:
`p->offset` got updated with `actual_unpack_bytes` in the line above.  So it points to the current offset.  We need the subtraction.

Reply:
I see. I wish we would modify the code to do the error checks first then update the states to avoid this confusions. Anyway, I am accepting it.

Reply:
Actually, the convoluted code is a result of we updated `p->in_data_sz` and `p->offset` with `actual_unpack_bytes` but we didn't update `payload_sz`. So if we add
```
payload_sz -= actual_unpack_bytes;
```
under line 192, I think the code will much easier to read.

Reply:
I don't think it'll be easier to read.   That'd not be the payload size anymore.